src/toil/provisioners/aws/awsProvisioner.py

```diff
@@ -0,0 +1,511 @@
+            profile_name = profile['arn'].split('/', 1)[1]
+            try:
+                ctx.iam.remove_role_from_instance_profile(profile_name, profile_name)
+            except BotoServerError as e:
+                if e.status == 404:
+                    pass
+                else:
+                    raise
+            ctx.iam.delete_instance_profile(profile_name)
+
+    def _addNodes(self, instancesToLaunch, preemptable=False):
+        bdm = self._getBlockDeviceMapping(self.instanceType)
+        arn = self._getProfileARN(self.ctx)
+        # quay.io/toil-leader:tag
+        workerData = dockerInfo.rsplit(':', 1)
+        workerRepo = workerData[0].rsplit('-', 1)[0] + '-worker'
+        workerTag = workerData[1]
+        workerData = {'role': 'worker', 'tag': workerTag,
+                      'args': workerArgs.format(ip=self.masterIP, preemptable=preemptable),
+                      'repo': workerRepo}
+        userData = AWSUserData.format(**workerData)
+        kwargs = {'key_name': self.keyName, 'security_groups': [self.clusterName],
+                  'instance_type': self.instanceType.name,
+                  'user_data': userData, 'block_device_map': bdm,
+                  'instance_profile_arn': arn}
+
+        if not preemptable:
+            logger.info('Launching %s non-preemptable master', instancesToLaunch)
+            create_ondemand_instances(self.ctx.ec2, image_id=coreOSAMI,
+                                      spec=kwargs, num_instances=1)
+        else:
+            logger.info('Launching %s preemptable nodes', instancesToLaunch)
+            # force generator to evaluate
+            list(create_spot_instances(ec2=self.ctx.ec2, price=self.spotBid, image_id=coreOSAMI,
+                                       clusterName=self.clusterName, spec=kwargs, num_instances=instancesToLaunch))
+        logger.info('Launched %s new instance(s)', instancesToLaunch)
+
+    @classmethod
+    def _getBlockDeviceMapping(cls, instanceType):
+        # determine number of ephemeral drives via cgcloud-lib
+        bdtKeys = ['', '/dev/xvdb', '/dev/xvdc', '/dev/xvdd']
+        bdm = BlockDeviceMapping()
+        # the first disk is already attached for us so start with 2nd.
+        for disk in xrange(1, instanceType.disks + 1):
+            bdm[bdtKeys[disk]] = BlockDeviceType(
+                ephemeral_name='ephemeral{}'.format(disk - 1))  # ephemeral counts start at 0
+
+        logger.debug('Device mapping: %s', bdm)
+        return bdm
+
+    def _removeNodes(self, instances, numNodes, preemptable=False, force=False):
+        # based off toil.provisioners.cgcloud.provisioner.CGCloudProvisioner._removeNodes()
+        logger.debug('Attempting to delete nodes - force = %s', force)
+        # If the batch system is scalable, we can use the number of currently running workers on
+        # each node as the primary criterion to select which nodes to terminate.
+        if isinstance(self.batchSystem, AbstractScalableBatchSystem):
```

Comment:
It's very similar to CGCloud but would require some changes in the CGCloud provisioner prior to factoring out the shared functionality. I'll open a ticket for this.